providers/implementations/exchange/ecdh_exch.c

```diff
@@ -0,0 +1,503 @@
+ * We happen to know that our KEYMGMT simply passes EC_KEY structures, so
+ * we use that here too.
+ */
+
+typedef struct {
+    OPENSSL_CTX *libctx;
+
+    EC_KEY *k;
+    EC_KEY *peerk;
+
+    /*
+     * ECDH cofactor mode:
+     *
+     *  * 0  disabled
+     *  * 1  enabled
+     *  * -1 default to cofactor mode set for *k
+     */
+    int cofactor_mode;
+
+    /************
+     * ECDH KDF *
+     ************/
+    /* KDF (if any) to use for ECDH */
+    enum kdf_type kdf_type;
+    /* Message digest to use for key derivation */
+    EVP_MD *kdf_md;
+    /* User key material */
+    unsigned char *kdf_ukm;
+    size_t kdf_ukmlen;
+    /* KDF output length */
+    size_t kdf_outlen;
+} PROV_ECDH_CTX;
+
+static
+void *ecdh_newctx(void *provctx)
+{
+    PROV_ECDH_CTX *pectx = OPENSSL_zalloc(sizeof(*pectx));
+
+    if (pectx == NULL)
+        return NULL;
+
+    pectx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    pectx->cofactor_mode = -1;
+    pectx->kdf_type = PROV_ECDH_KDF_NONE;
+
+    return (void*)pectx;
+}
+
+static
+int ecdh_init(void *vpecdhctx, void *vecdh)
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+
+    if (pecdhctx == NULL || vecdh == NULL || !EC_KEY_up_ref(vecdh))
+        return 0;
+    EC_KEY_free(pecdhctx->k);
+    pecdhctx->k = vecdh;
+    pecdhctx->cofactor_mode = -1;
+    pecdhctx->kdf_type = PROV_ECDH_KDF_NONE;
+    return 1;
+}
+
+static
+int ecdh_set_peer(void *vpecdhctx, void *vecdh)
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+
+    if (pecdhctx == NULL || vecdh == NULL || !EC_KEY_up_ref(vecdh))
+        return 0;
+    EC_KEY_free(pecdhctx->peerk);
+    pecdhctx->peerk = vecdh;
+    return 1;
+}
+
+static
+void ecdh_freectx(void *vpecdhctx)
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+
+    EC_KEY_free(pecdhctx->k);
+    EC_KEY_free(pecdhctx->peerk);
+
+    EVP_MD_free(pecdhctx->kdf_md);
+    OPENSSL_clear_free(pecdhctx->kdf_ukm, pecdhctx->kdf_ukmlen);
+
+    OPENSSL_free(pecdhctx);
+}
+
+static
+void *ecdh_dupctx(void *vpecdhctx)
+{
+    PROV_ECDH_CTX *srcctx = (PROV_ECDH_CTX *)vpecdhctx;
+    PROV_ECDH_CTX *dstctx;
+    unsigned char *tmp_kdf_ukm = NULL;
+
+    dstctx = OPENSSL_zalloc(sizeof(*srcctx));
+    if (dstctx == NULL)
+        return NULL;
+
+    if (srcctx->k != NULL && !EC_KEY_up_ref(srcctx->k))
+        goto err;
+    else
+        dstctx->k = srcctx->k;
+
+    if (srcctx->peerk != NULL && !EC_KEY_up_ref(srcctx->peerk))
+        goto err;
+    else
+        dstctx->peerk = srcctx->peerk;
+
+    if (srcctx->kdf_md != NULL && !EVP_MD_up_ref(srcctx->kdf_md))
+        goto err;
+    else
+        dstctx->kdf_md = srcctx->kdf_md;
+
+    if (srcctx->kdf_ukm != NULL
+            && (tmp_kdf_ukm = OPENSSL_memdup(srcctx->kdf_ukm,
+                                             srcctx->kdf_ukmlen)) == NULL) {
+        goto err;
+    } else {
+        dstctx->kdf_ukm = tmp_kdf_ukm;
+    }
+
+    *dstctx = *srcctx;
```

Comment:
This could be moved above the memdup, meaning we wouldn't have to fix up `kdf_ukm`

Reply:
I did this on purpose to avoid even temporarily having `dstctx->kdf_ukm` pointing to the same memory address of `srcctx->kdf_ukm`!

My thinking is that given our usual `goto err;` pattern, we should avoid `kdf_ukm` ever being in a invalid state as in the future, maintaining this code, there is a concrete risk of exiting early with an error and having around an object that could lead to double frees.

Reply:
Something we've done previously is do the assignment at the top, and then immediately set all ptrs in the dest object to be NULL. Then the rest of the function goes through and sets the correct ptrs. If anything fails halfway through then the invalid ptrs are already NULL.

Reply:
Fixed in upcoming push